### DISCUSSION TITLE
Pub/Sub Example?

### DISCUSSION THREAD

**Discussion by johnnytolengo** [cirospaciari/socketify.py#46] [Q&A] [answered]
Hello, it seems that this lib supports pub/sub but I have not clue how to handle it?
is it there some example?

**Comment by cirospaciari** [comment, marked as answer]
Hello @johnnytolengo we have some examples in
https://github.com/cirospaciari/socketify.py/tree/main/examples 
Chat app example use pub/sub 

And also in:
https://github.com/cirospaciari/socketify.py/tree/main/bench/websockets

See the example bellow as used in benchmark, you can use ws.subscribe(topic) and ws.unsusbscribe(topic) and use app.publish(topic, message, opcode) to broadcasting the message to the subscribed clients

```python
 from socketify import App, AppOptions, OpCode, CompressOptions

remaining_clients = 16
app = App(websocket_factory_max_itens=1_500_000)


def ws_open(ws):
    ws.subscribe("room")
    global remaining_clients
    remaining_clients = remaining_clients - 1
    if remaining_clients == 0:
        print("All clients connected")
        print('Starting benchmark by sending "ready" message')

        app.publish("room", "ready", OpCode.TEXT)


def ws_message(ws, message, opcode):
    app.publish("room", message, opcode)

def ws_close(ws, close, message):
    global remaining_clients
    remaining_clients = remaining_clients + 1


app.ws(
    "/*",
    {
        "compression": CompressOptions.DISABLED,
        "max_payload_length": 16 * 1024 * 1024,
        "idle_timeout": 60,
        "open": ws_open,
        "message": ws_message,
        "close": ws_close,
    },
)
app.any("/", lambda res, req: res.end("Nothing to see here!'"))
app.listen(
    4001,
    lambda config: print("Listening on port http://localhost:%d now\n" % (config.port)),
)
app.run()
```

**Comment by johnnytolengo** [comment]
That's fine, now looking for the message dataStructure (protocol)  for that pub/sub on both sides client/server. I assume that info is in the uWebsockets documentation including internal api's to get connected clients, stats, etc, etc...

Thanks

**Reply by cirospaciari** [reply]
Uses default WebSockets protocol check the chat example, you just send an message from the client side and subscribe in ws_message()  we implemented all features from uWebSockets and a lot more, you can use app
num_subscribers(topic) to get how many people are there and use per socket data too like in chat example. If you need anything just asks here that I will help you, you can also check our discord server https://discord.socketify.dev/

**Comment by johnnytolengo** [comment]
It's interesting how the Server handle(eg. more than 1 connection with the same ClientID) the subscriptions(multiple), it supposed to be something very similar to how MQTT brokers do. 
I am now join the discord channel.
Thank you
